luarules/gadgets/game_initial_spawn.lua

```diff
@@ -43,30 +43,96 @@ if gadgetHandler:IsSyncedCode() then
 	-- Vars
 	----------------------------------------------------------------
 	local validStartUnits = {}
-	local armcomDefID = UnitDefNames.armcom and UnitDefNames.armcom.id
-	if armcomDefID then
-		validStartUnits[#validStartUnits+1] = armcomDefID
-	end
-	local corcomDefID = UnitDefNames.corcom and UnitDefNames.corcom.id
-	if corcomDefID then
-		validStartUnits[#validStartUnits+1] = corcomDefID
-	end
-	local legcomDefID = UnitDefNames.legcom and UnitDefNames.legcom.id
-	if legcomDefID then
-		validStartUnits[#validStartUnits+1] = legcomDefID
-	end
 
 	local RANDOM_DUMMY = UnitDefNames.dummycom and UnitDefNames.dummycom.id
 
-	local function isUnitValid(unitDefID)
-		if not unitDefID then
-			return false
-		end
-		if table.contains(validStartUnits, unitDefID) then
-			return true
-		end
-		if unitDefID == RANDOM_DUMMY then
-			return true
+	local getValidRandom, isUnitValid
+
+	do
+		local modoptions = Spring.GetModOptions()
+		local factionlimiter = tonumber(modoptions.factionlimiter) or 0
+		if factionlimiter > 0 then
+			local legcomDefID = UnitDefNames.legcom and UnitDefNames.legcom.id
+			local armcomDefID = UnitDefNames.armcom and UnitDefNames.armcom.id
+			local corcomDefID = UnitDefNames.corcom and UnitDefNames.corcom.id
+
+			local allyTeams = Spring.GetAllyTeamList()
+			for i = 1, #allyTeams do
+				local allyTeam = allyTeams[i]
+				local allyStartUnits = {}
+				-- for some reason lua ain't liking using #allyStartUnits
+				local unitsCount = 1
+
+				local allyTeamBitmask = math.bit_and(math.floor(factionlimiter/2^(allyTeam*3)), 7)
+				allyTeamBitmask = allyTeamBitmask == 0 and 7 or allyTeamBitmask
+
+				if legcomDefID then
+					if math.bit_and(allyTeamBitmask, 4) == 4 then
```

Comment:
if we're sticking with bitmasks it would be good to have named constants like
```lua
local ARM_MASK = 2^0
local COR_MASK = 2^1
local LEG_MASK = 2^2
```